### ISSUE TITLE
Named tuples in import cycles still fail sometimes

### ISSUE THREAD

**Issue by ilevkivskyi** [python/mypy#4442]
It looks like the issue #3054 was not completely fixed. Namely, in some scenarios named tuples
in import cycles result in ``Invalid type``. For example, checking this file
```python
import posix
```
results in
```
typeshed/stdlib/3/os/__init__.pyi:331: error: Invalid type "posix.uname_result"
typeshed/stdlib/3/os/__init__.pyi:599: error: Invalid type "posix.times_result"
typeshed/stdlib/3/os/__init__.pyi:605: error: Invalid type "posix.waitid_result"
typeshed/stdlib/3/os/__init__.pyi:622: error: Invalid type "posix.sched_param"
typeshed/stdlib/3/os/__init__.pyi:624: error: Invalid type "posix.sched_param"
typeshed/stdlib/3/os/__init__.pyi:625: error: Invalid type "posix.sched_param"
```
Note that the above "invalid" types are all named tuples and there is an import cycle ``os <-> posix``. Interestingly, ``import os`` and ``from os import *`` both typecheck cleanly.

I haven't yet tried to find a simple repro (not involving typrshed).

**Comment by ilevkivskyi** [comment]
Somehow this breaks typshed build since recently, so I raised priority to high.

**Comment by elliott-beach** [comment]
I deleted code from the typestub until I got, in `posix.pyi`,

```python
from typing import NamedTuple
from os import stat_result

# This was NamedTuple('etc'), but that doesn't affect the error.
uname_result = 5 # E: Invalid type "posix.uname_result" 
```

and in `os/__init__.pyi`,
```python
stat_result = "dummy"
from posix import T
def uname() -> uname_result: ...
```
so there is something fishy with the way circular imports are resolved, but it doesn't seem to me to involve NamedTuple specifically.

**Comment by elliott-beach** [comment]
Actually, I think this is just a case of the circular import problem discussed in https://stackoverflow.com/a/40094439/5749914 (posix.py is Y.py, os/\_\_init__.py is X.py).

**Comment by gvanrossum** [comment]
Do note that circular dependencies at run-time work differently than in
mypy's analysis.

**Comment by elliott-beach** [comment]
Yeah, but if it wouldn't work by python's runtime I figure there is no reason to expect it to work for mypy.

**Comment by elliott-beach** [comment]
@ilevkivskyi I think this ought to be closed because the behavior is reasonable.

**Comment by JukkaL** [comment]
@ilevkivskyi Can you come up with a self-contained example that reproduces the problem? I'd like to know if #4695 fixed this issue, and if not, I'll look at fixing this.

**Comment by ilevkivskyi** [comment]
@JukkaL Unfortunately, I didn't yet find a simple repro, but there is a "dirty" way of testing it by just using the offending typeshed commit `81751306883`. I just tried it, and the error is still present, it was not fixed by #4695.

**Comment by ilevkivskyi** [comment]
@JukkaL 
> and if not, I'll look at fixing this.

OK, I will spend some more time now to find a simple repro for you.

**Comment by ilevkivskyi** [comment]
So, here is a simplest repro I can find:
```python
# file a.py
class C:
    pass

from b import tp
x: tp

# file b.py
from a import C
from typing import NamedTuple

tp = NamedTuple('tp', [('x', int)])
```
Then one needs to run `mypy b.py` and get:
```
a.py:5: error: Invalid type "b.tp"
```
Note that error disappears if `tp` is a normal class.

Also note that the same error also appears if one or both of `a` and `b` are packages and use named tuples in `__init__.py`.
